client/telemetry/src/lib.rs

```diff
@@ -16,339 +16,486 @@
-	// Build the list of telemetry endpoints.
-	let (endpoints, wasm_external_transport) = (config.endpoints.0, config.wasm_external_transport);
-
-	let (sender, receiver) = mpsc::channel(16);
-	let guard = {
-		let logger = TelemetryDrain { sender: std::panic::AssertUnwindSafe(sender) };
-		let root = slog::Logger::root(slog::Drain::fuse(logger), slog::o!());
-		slog_scope::set_global_logger(root)
-	};
-
-	let worker = match worker::TelemetryWorker::new(endpoints, wasm_external_transport) {
-		Ok(w) => Some(w),
-		Err(err) => {
-			error!(target: "telemetry", "Failed to initialize telemetry worker: {:?}", err);
-			None
+impl TelemetryWorker {
+	pub(crate) fn new(buffer_size: usize, transport: WsTrans) -> Self {
+		let (message_sender, message_receiver) = mpsc::channel(buffer_size);
+		let (register_sender, register_receiver) = mpsc::unbounded();
+
+		Self {
+			message_receiver,
+			message_sender,
+			register_receiver,
+			register_sender,
+			transport,
 		}
-	};
+	}
 
-	Telemetry {
-		inner: Arc::new(Mutex::new(TelemetryInner {
-			worker,
-			receiver,
-		})),
-		_guard: Arc::new(guard),
+	/// Get a new [`TelemetryHandle`].
+	///
+	/// This is used when you want to register a new telemetry for a Substrate node.
+	pub fn handle(&self) -> TelemetryHandle {
+		TelemetryHandle {
+			message_sender: self.register_sender.clone(),
+		}
 	}
-}
 
-/// Event generated when polling the worker.
-#[derive(Debug)]
-pub enum TelemetryEvent {
-	/// We have established a connection to one of the telemetry endpoint, either for the first
-	/// time or after having been disconnected earlier.
-	Connected,
-}
+	/// Get a clone of the channel's `Sender` used to send telemetry events.
+	pub(crate) fn message_sender(&self) -> mpsc::Sender<TelemetryMessage> {
+		self.message_sender.clone()
+	}
 
-impl Stream for Telemetry {
-	type Item = TelemetryEvent;
-
-	fn poll_next(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Option<Self::Item>> {
-		let before = Instant::now();
-
-		// Because the `Telemetry` is cloneable, we need to put the actual fields behind a `Mutex`.
-		// However, the user is only ever supposed to poll from one instance of `Telemetry`, while
-		// the other instances are used only for RAII purposes.
-		// We assume that the user is following this advice and therefore that the `Mutex` is only
-		// ever locked once at a time.
-		let mut inner = match self.inner.try_lock() {
-			Some(l) => l,
-			None => {
-				warn!(
+	/// Run the telemetry worker.
+	///
+	/// This should be run in a background task.
+	pub async fn run(self) {
+		let Self {
+			mut message_receiver,
+			message_sender: _,
+			mut register_receiver,
+			register_sender: _,
+			transport,
+		} = self;
+
+		let mut node_map: HashMap<Id, Vec<(u8, Multiaddr)>> = HashMap::new();
+		let mut node_pool: HashMap<Multiaddr, _> = HashMap::new();
+
+		// initialize the telemetry nodes and register new notifiers
+		async fn process_register(
```

Comment:
Perfect :+1: 

Now just move this and the other function below out of the `run` function :P

Reply:
Done in 24596be78

(...like that??)